File: Code Chef/SEVENSEG.cpp
#include<bits/stdc++.h>
 
using namespace std;
 
int mask[10]={119, 36, 93, 109, 46, 107, 123, 37, 127, 111};
int n;
int x[11], y[11];
 
int check(int msk)
{
    for (int i=1;i<=n;i++)
    {
	int bed = 0;
	for (int j=0;j<7;j++)
	if ((mask[x[i]]&(1<<j)) && !(msk&(1<<j))) bed++;
	if (bed!=y[i]) return false;
    }
 
     return true;
}
void solve()
{
    assert(n<=10);
    cin>>n;
 
    for (int i=1;i<=n;i++){
	cin>>x[i]>>y[i];
	assert(x[i]>=0 && x[i]<=9);
	assert(y[i]<=7 && y[i]>=0);
 
	for (int j=i-1;j>0;j--)
	    assert(x[i]!=x[j]);
    }
 
    int ok = 0;
    int mini = 10;
    int maxi = 0;
    for (int i=0;i<1<<7;i++)
	if (check(i)){
	    if (!ok)
	    {
	        ok = 1;
	        mini = __builtin_popcount(i);
	        maxi = mini;
	    } else
	    {
	        mini = min(mini, __builtin_popcount(i));
	        maxi = max(maxi, __builtin_popcount(i));
	    }
	}
 
    if (!ok) printf("invalid\n"); else
    printf("%d %d\n",mini,maxi);
}
 
int main()
{
    int t;
 
    cin>>t;
    while(t--)
	solve();
 
    return 0;
}

File: Code Chef/SEVENSEG.py
def  countSetBits(n): 
	count = 0
	while (n): 
		count += n & 1
		n >>= 1
	return count

for _ in range(int(input())):
	la = []
	lb = []
	masks = [63,6,91,79,102,109,125,7,127,111]
	conf = []
	for o in range(int(input())):
		n,k = list(map(int,input().split()))
		la.append(n)
		lb.append(k)

	q = len(la)
	for i in range(128):
		f = 1
		for j in range(q):
			if countSetBits(masks[la[j]]&i) != lb[j]:
				f = 0
				break
		if f:
			conf.append(i)

	k = max(lb)
	mi =999
	ma = -999
	if len(conf)>0:
		for u in conf:
			p = 7 - countSetBits(u)
			mi = min(mi,p)
			ma = max(ma,p)
		print(mi,ma)
	else:
		print('invalid')
 